game/src/main/java/org/apollo/game/sync/task/PlayerSynchronizationTask.java

```diff
@@ -1,171 +1,772 @@
-import org.apollo.game.sync.block.ChatBlock;
-import org.apollo.game.sync.block.SynchronizationBlock;
-import org.apollo.game.sync.block.SynchronizationBlockSet;
-import org.apollo.game.sync.seg.*;
-
-import java.util.ArrayList;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Set;
-import java.util.stream.Stream;
+import org.apollo.game.model.entity.PlayerUpdateInfo;
+import org.apollo.game.model.entity.setting.Gender;
+import org.apollo.game.model.inv.Inventory;
+import org.apollo.game.sync.block.*;
+import org.apollo.game.sync.block.TitleBlock;
+import org.apollo.net.codec.game.DataOrder;
+import org.apollo.net.codec.game.DataTransformation;
+import org.apollo.net.codec.game.DataType;
+import org.apollo.net.codec.game.GamePacketBuilder;
+import org.apollo.net.meta.PacketType;
+import org.apollo.util.BufferUtil;
+import org.jetbrains.annotations.NotNull;
+
+import static org.apollo.game.model.entity.PlayerUpdateInfo.MAXIMUM_LOCAL_PLAYERS;
 
 /**
  * A {@link SynchronizationTask} which synchronizes the specified {@link Player} .
  *
- * @author Graham
+ * @author Kris
+ * @author Dragonkk
+ * @author Cjay0091
  */
 public final class PlayerSynchronizationTask extends SynchronizationTask {
 
+	private static final int[][] PACKED_LOCATION_UPDATE2 = new int[][]{{0, 3, 5}, {1, -1, 6}, {2, 4, 7}};
+
+	private final Player player;
+	private final PlayerUpdateInfo info;
+	private final World world;
+	private final ObjectSet<Player> limitedPlayers;
+
+	private ObjectOpenHashSet<Player> added;
+
+	public PlayerSynchronizationTask(Player player) {
+		this.player = player;
+		this.info = player.getUpdateInfo();
+		this.world = player.getWorld();
+
+		this.limitedPlayers = player.getLocalPlayerList();
+		this.added = new ObjectOpenHashSet<>();
+	}
+
+	@Override
+	public void run() {
+		info.buffer = new GamePacketBuilder(79, PacketType.VARIABLE_SHORT);
+		info.largeMaskBuffer = new GamePacketBuilder();
+		limitedPlayers.clear();
+		prefetch();
+		processLocalPlayers(true);
+		processLocalPlayers(false);
+		processOutsidePlayers(true);
+		processOutsidePlayers(false);
+		info.buffer.putBytes(info.largeMaskBuffer);
+		info.localIndexesCount = 0;
+		info.externalIndexesCount = 0;
+		for (var playerIndex = 1; playerIndex < 2048; playerIndex++) {
+			info.activityFlags[playerIndex] >>= 1;
+			if (info.localPlayers[playerIndex] == null) {
+				info.externalIndexes[info.externalIndexesCount++] = playerIndex;
+			} else {
+				info.localIndexes[info.localIndexesCount++] = playerIndex;
+			}
+		}
+		player.send(new PlayerSynchronizationMessage(info.buffer));
+	}
+
 	/**
-	 * The maximum amount of local players.
+	 * Prefetches a list of players closest to us when there are more than 255 players in the viewport.
+	 * TODO: Convert the LocationMap to function at the chunk level. No reason to separate them.
 	 */
-	private static final int MAXIMUM_LOCAL_PLAYERS = 255;
+	private void prefetch() {
+		limitedPlayers.clear();
```

Comment:
It looks like this is an approximation of:

```
for (Player updatingPlayer : playersToUpdate) {
   for (Region region : regions) {
      for (Player candidate: regions.getPlayers()) {
         // test
      }
   }
}
```

is that right?

Reply:
Hm, this is what we do currently. That doesn't quite scale. @Major- @rmcmk ever thought about this? Scanning close regions or adding to this change buffer with an event system seems like a better idea.

Reply:
That's not exactly correct. 

 for (Region region : regions) {
      for (Player candidate: regions.getPlayers()) {
         // test
      }
   }

Is more what it is like in that function. If you're referring to what it does for every player combined,, then yes.